Add Broadcom bnxt pta and drivers

shitalt opened the pull request:
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

sandeepbrcm commented on core/pta/sub.mk:
```diff
@@ -9,3 +9,5 @@ srcs-$(CFG_SECSTOR_TA_MGMT_PTA) += secstor_ta_mgmt.c
 srcs-$(CFG_WITH_STATS) += stats.c
 srcs-$(CFG_SYSTEM_PTA) += system.c
+
+subdirs-y += bcm_pta
```
Thanks @jforissier .  We have a few more Broadcom specific ptas which possibly can be up streamed in future. So do you suggest to create a sub-dir in that case. I thought this way we can segregate the soc-specific ones from the generic ones in core/pta/*.

jenswi-linaro replied:
A `bcm` directory here makes sense, we can drop the `_pta` part since it's redundant.

jforissier replied:
@sandeepbrcm @jenswi-linaro agreed

shitalt replied:
Ok. Will update the folder name to `bcm`

jenswi-linaro commented on core/pta/bcm_pta/sub.mk:
```diff
@@ -0,0 +1,2 @@
+global-incdirs-y += include
```
I don't suppose any C files will use `pta_bcm_*.h` files but files in/under  this directly.
How about putting everything directly in this directory with no further hierarchy (skipping the separate `src` directory too)?

shitalt replied:
Agreed.

jenswi-linaro commented on core/drivers/bnxt/bnxt.c:
```diff
@@ -0,0 +1,189 @@
+
+static TEE_Result bnxt_init(void)
+{
+	bnxt_access_window_virt_addr = (vaddr_t)phys_to_virt
```
Regardless how you break the line, the `(` should follow directly after the function name. How about something like this:
``` C
        bnxt_access_window_virt_addr =
                (vaddr_t)phys_to_virt(NIC400_BNXT_IDM_IO_CONTROL_DIRECT,
                                      MEM_AREA_IO_SEC);
        bnxt_indirect_dest_addr = (vaddr_t)phys_to_virt(BNXT_INDIRECT_BASE,
                                                        MEM_AREA_IO_SEC);
```
instead?

shitalt replied:
Ok, will re-align as suggested.

jenswi-linaro commented on core/drivers/bnxt/bnxt_fw.c:
```diff
@@ -0,0 +1,684 @@
+	struct ver_ext_offset_s	version2;
+};
+
+#pragma pack(1)
```
I don't know what this means, but I suppose it has something to do with the __packed attribute which we have. Could that be used where needed instead?

sandeepbrcm replied:
@shitalt 
#pragma pack(1)
  xxx
#pragma pack()
can be replaced by 
 struct ape_bin_hdr_s {
 xxx
} __packed;
 or __packed  __aligned(1);    but for sure test the output.

jenswi-linaro commented on core/drivers/bnxt/bnxt_fw.c:
```diff
@@ -0,0 +1,684 @@
+				     uint8_t *outbuf,
+				     size_t outbuf_size)
+{
+	int i, j, k, r, c;
```
All local variables should normally be initialized with a value. The rationale is that in case some variable isn't assigned another value later we still have a rather defined behaviour and we're not giving away junk from the stack.

I'm only commenting in this function, but the comment applies to the entire pull request.

shitalt replied:
Will have initialized local variables here and in other functions

shitalt commented on core/arch/arm/plat-bcm/conf.mk:
```diff
@@ -27,6 +27,8 @@ $(call force,CFG_SP805_WDT,y)
 $(call force,CFG_BCM_SOTP,y)
 $(call force,CFG_BCM_GPIO,y)
+# enable BNXT fw load driver
+$(call force,CFG_BNXT_FW,y)
```
1. Ok, will remove the comment
2. Will update

shitalt commented on core/drivers/bnxt/bnxt.c:
```diff
@@ -0,0 +1,188 @@
+ * Copyright 2019 Broadcom.
+ */
+
+#include <drivers/bcm-include/bnxt.h>
```
Ok.

shitalt commented on core/drivers/bnxt/bnxt.c:
```diff
@@ -0,0 +1,188 @@
+	if (num_entries == 0)
+		return 0;
+
+	/* only write up to the next window boundary */
```
ok. will check other comments as well

shitalt commented on core/drivers/bnxt/bnxt_fw.c:
```diff
@@ -0,0 +1,684 @@
+	uint8_t num_loadable_sections;
+	uint32_t checksum;
+};
+
```
Done. removing this.

shitalt commented on core/drivers/bnxt/bnxt_fw.c:
```diff
@@ -0,0 +1,684 @@
+	struct ape_mem_region_s *region = ape_mem_regions;
+	uint32_t addr = 0;
+
+	for ( ; region->size != 0; region++) {
```
ok

shitalt commented on core/include/drivers/bcm-include/crc32.h:
```diff
@@ -0,0 +1,93 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
```
Agreed but will wait until it is used by others

shitalt commented on core/pta/bcm_pta/include/pta_bnxt.h:
```diff
@@ -0,0 +1,18 @@
+	PTA_BNXT_FASTBOOT = 0,
+} pta_bnxt_cmd;
+
+#define BNXT_TA_NAME		"pta_bnxt.ta"
```
Done, merge the header with C file and will remove the "include" and "src" folder as suggested in other comments

shitalt commented on core/pta/bcm_pta/src/sub.mk:
```diff
@@ -0,0 +1 @@
+srcs-$(CFG_BNXT_FW) += pta_bnxt.c
```
I will accommodate the comment

shitalt commented on core/drivers/bnxt/bnxt_fw.c:
```diff
@@ -0,0 +1,684 @@
+			     section->org_data_len, size);
+			goto ape_section_copy_exit;
+		}
+		src = (uintptr_t)(work_buff);
```
Done

shitalt commented on core/drivers/bnxt/bnxt_fw.c:
```diff
@@ -0,0 +1,684 @@
+		checksum = CRC32(CRC32_INIT_VAL, (const char *)src, size);
+	} else {
+		checksum = 0;
+		for (i = 0; i < (size / sizeof(uint32_t)); i++)
```
Done

shitalt commented on core/drivers/bnxt/bnxt_fw.c:
```diff
@@ -0,0 +1,684 @@
+	 * First we write into local memory to calculate CRC before
+	 * updating into Nitro memory
+	 */
+	info = (struct nitro_crash_addr_info *)malloc
```
Ok. Will remove the cast and re-align the code.

shitalt commented on core/drivers/bnxt/bnxt_fw.c:
```diff
@@ -0,0 +1,684 @@
+	info = (struct nitro_crash_addr_info *)malloc
+					(sizeof(struct nitro_crash_addr_info));
+	if (!info) {
+		EMSG("ERROR: buffer allocation\n");
```
Done

shitalt commented on core/pta/bcm_pta/include/pta_bnxt.h:
```diff
@@ -0,0 +1,18 @@
+		{0x6272636D, 0x2019, 0x0716,  \
+		{0x42, 0x43, 0x4D, 0x5F, 0x53, 0x43, 0x48, 0x49} }
+
+enum {
```
ok will include the suggestion

shitalt commented on core/pta/bcm_pta/src/pta_bnxt.c:
```diff
@@ -0,0 +1,70 @@
+	DMSG("destroy entry point for static ta \"%s\"", BNXT_TA_NAME);
+}
+
+static TEE_Result open_session(uint32_t nParamTypes __unused,
```
Ok and will update at other places too.

jforissier commented:
`Acked-by: Jerome Forissier <jerome@forissier.org>`

sandeepbrcm commented on core/arch/arm/plat-bcm/crc32.h:
```diff
@@ -0,0 +1,18 @@
+#define CRC32_INIT_VAL	(~0)
+#define CRC32		crc32i
+
+uint32_t ucrc32(const char ch, uint32_t crc);
```
`ucrc32()`  not to be exported. Make it static in  C file and remove from here.

shitalt replied:
Done. Made ucrc32 static function.

sandeepbrcm commented:
`Reviewed-by: Sandeep Tripathy <sandeep.tripathy@broadcom.com>`

shitalt commented:
Hi @jenswi-linaro 

I've addressed the review comments in the pull request.
Please let me know if you any other comments or if this pull request is good.

shitalt commented on core/drivers/bnxt/bnxt.c:
```diff
@@ -0,0 +1,189 @@
+	for (i = 0; i < num_entries; i++) {
+		io_write32(target, *(uint32_t *)src);
+		target += sizeof(uint32_t);
+		src += (src_4byte_increment) ? sizeof(uint32_t) : 0;
```
Will include the suggestion.

shitalt commented on core/arch/arm/plat-bcm/crc32.c:
```diff
@@ -0,0 +1,88 @@
+
+static uint32_t ucrc32(const char ch, uint32_t crc)
+{
+	return (crc32tbl[(crc ^ ch) & 0xff] ^ (crc >> 8));
```
Ok. will remove it

jforissier commented:
@shitalt since all the comments have been addressed now, would you please squash the fixups where they belong and add the Review/Ack tags? Thanks!

jforissier commented:
@shitalt thank you!
